Tribler/community/tunnel/tunnel_community.py

```diff
@@ -697,19 +745,33 @@ def _ours_on_created_extended(self, circuit, message):
         circuit.unverified_hop = None
 
         if circuit.state == CIRCUIT_STATE_EXTENDING:
+            become_exit = circuit.goal_hops - 1 == len(circuit.hops)
+            ignore_candidates = [self.crypto.key_to_bin(hop.public_key) for hop in circuit.hops] + \
+                [self.my_member.public_key]
 
-            if circuit.goal_hops - 1 == len(circuit.hops) and circuit.required_exit:
-                host, port, pub_key = circuit.required_exit
-                extend_hop_public_bin = pub_key
-                extend_hop_addr = (host, port)
+            if become_exit:
+                if circuit.required_exit:
+                    # Set the required exit according to the circuit setting (e.g. for linking e2e circuits)
+                    host, port, pub_key = circuit.required_exit
+                    extend_hop_public_bin = pub_key
+                    extend_hop_addr = (host, port)
+                else:
+                    # The exit node is a verified connectable exit node peer chosen by the circuit initiator
+                    extend_hop_public_bin = None
+                    for c in self.dispersy_yield_verified_candidates():
+                        pubkey = c.get_member().public_key
+                        exit_candidate = self.exit_candidates[pubkey]
+                        if exit_candidate.become_exit and self.candidate_is_connectable(c) and pubkey not in ignore_candidates:
+                            wan = c._wan_address
+                            extend_hop_public_bin = pubkey
+                            extend_hop_addr = wan
+                            break
```

Comment:
I would move this logic into the create_circuit method. Then if a required_exit is passed, do nothing, but if the required_exit is None then look for an exit-node in the verified candidates.
This will ensure that we never attempt to create a circuit without having an exit-node.